### ISSUE TITLE
Leakage Detection on sdewac-v3

### ISSUE THREAD

**Issue by divyansh-kaushik1** [CastIehard/LLM-Grammaticalization-Prediction#14]
# Leakage Check: sdewac-v3.tagged using TinyLLaMA and Flan-T5

## Objective

To test whether modern language models (TinyLLaMA and Flan-T5) show signs of training data contamination or leakage by reproducing sentences from the **sdewac-v3.tagged** corpus.

---

## Methodology

The methodology implements a simplified version of the **VERI detection** strategy for the paper *“Leak, Cheat, Repeat”* (arXiv:2402.03927v2):

- We **randomly sample sentences** from the `sdewac-v3.tagged` corpus.
- For each sentence:
  - Extract the **first N lines** as a prompt (default: 4).
  - Prompt the model to **continue the POS-tagged sentence**.
  - Measure overlap between the generated output and the original sentence.

Two overlap scores are calculated:
- `ref_overlap`: % of reference lines present in the model’s output.
- `out_overlap`: % of output lines that appear in the reference.

A generation is considered a **leak** if:
```
ref_overlap ≥ 0.8 or out_overlap ≥ 0.8
```

---

### Common Code
- `parse_tagged_file()` reads the `.tagged` file into structured sentence blocks.

### TinyLLaMA Model
- Runs using [`llama-cpp-python`](https://github.com/ggerganov/llama.cpp) and a local `.gguf` quantized model.
- Inference uses `Llama()` from `llama_cpp`.

### Flan-T5 Model
- Runs using HuggingFace Transformers (`AutoModelForSeq2SeqLM`).
- Inference on GPU (if available), using default `generate()` pipeline.

Both scripts write results as `.csv` with columns:
```
index, model, prompt, output, ref_overlap, out_overlap, leaked
```

---

## Alignment with the Paper

While this work is inspired by the **motivations** in *“Leak, Cheat, Repeat: Data Contamination and Evaluation Malpractices in Closed-Source LLMs” (arXiv:2402.03927)*, it differs in **methodological approach**.

> The paper uses a **literature review** to trace benchmark contamination, while this project takes an **empirical prompting approach** to detect **verbatim memorization**, aligning more with **training data extraction** research.

| Aspect                        | Paper: “Leak, Cheat, Repeat”                              | This Implementation                                       |
|-------------------------------|------------------------------------------------------------|------------------------------------------------------------|
| Goal                          | Systematic review of evaluation and benchmark leakage      | Empirical detection of verbatim overlap from test samples  |
| Target                        | ChatGPT (GPT-3.5/4) + research papers using it             | Flan-T5 and TinyLLaMA (open-source models)                |
| Evidence found                | Yes — benchmark reuse and indirect data leaks             | No — no leakage found using prefix-based regeneration      |

---

## Results & Analysis

After running the leakage detection pipeline across 10,000 sentences from the `sdewac-v3.tagged` corpus using **Flan-T5** and **TinyLLaMA**, here are the main findings:

---

### Leakage Summary Table

Both models showed **0% leakage** under the defined threshold condition (`overlap ≥ 0.8`):

![Leak Summary Table](https://github.com/user-attachments/assets/5ba1e866-53d3-4a6b-a8f9-c03a1bd8d446)

---

### Overlap Distributions

**Flan-T5** consistently showed *zero* reference or output overlaps across all samples, indicating abstract, non-memorized completions.  
**TinyLLaMA**, while not leaking, had some overlap variability suggesting potential structural mimicry without direct copying.

![Boxplot Overlaps](https://github.com/user-attachments/assets/19e67fd7-3a66-4e5c-a712-6caedc24f944)

---

### Score Histograms for TinyLLaMA

TinyLLaMA’s overlaps are skewed toward low values with occasional spikes in `out_overlap`.

![TinyLLaMA Histograms](https://github.com/user-attachments/assets/e946707d-0f48-4388-8890-e9d650154306)

---

### Reference vs Output Scatter (All Models)

No clusters reach the leak zone (`>0.8 overlap`), but TinyLLaMA has isolated points of medium similarity.

![Leakage Scatterplot](https://github.com/user-attachments/assets/b85de196-8aab-4f3d-ab4c-5f50039f0a26)

---

### Heatmap of Total Overlap Scores

Sum of overlap scores by model shows that **TinyLLaMA** accumulates slightly more overlap than **Flan-T5**, even though both stay under the leak threshold.

## Leakage Heatmap

![Leakage Heatmap](https://github.com/user-attachments/assets/bba3f790-01d1-4dad-bdc3-901588bd993a)

---

### Top 5 High-Overlap Prompts (Non-Leaked)

Examples with highest overlap that *did not* cross the leak threshold still reveal how TinyLLaMA can replicate structure more closely than Flan-T5.

![Top Prompts Table](https://github.com/user-attachments/assets/0428559d-6724-48b4-99ec-601e83ecb542)

---

### Conclusion

- **Flan-T5** demonstrates minimal verbatim copying risk — good generalization.
- **TinyLLaMA** stays below leakage thresholds but may reproduce certain sentence structures or syntactic patterns more rigidly.
- No actual contamination from the `sdewac-v3.tagged` corpus was detected.

---

### Papers
- Huang, Potts (2024). [Demystifying Verbatim Memorization in LLMs](https://aclanthology.org/2024.emnlp-main.598.pdf) — Core methodology for detecting verbatim recall using structured prompts.
- Carlini et al. (2021). [Extracting Training Data from Large Language Models (arXiv:2012.07805)](https://arxiv.org/pdf/2012.07805) — Seminal paper demonstrating real-world data leakage from GPT-2 using sampling + filtering.
- Djiré et al. (2025). *Detecting Memorization via Input Perturbation (PEARL)* — Black-box strategy to differentiate memorization from generalization.
- Reference Paper: [Leak, Cheat, Repeat (arXiv:2402.03927)](https://arxiv.org/abs/2402.03927) — Motivates the need to audit benchmarks for contamination, although via literature review rather than probing.

### Dataset
- `sdewac-v3.tagged` — German POS-tagged web corpus used as the target for leakage evaluation.

### Models Evaluated
- `google/flan-t5-base` — Open-source sequence-to-sequence model fine-tuned on instruction tasks.
- `TheBloke/TinyLlama-1.1B-Chat-v1.0-GGUF` — Quantized TinyLLaMA model for efficient inference using `llama-cpp`.

### Related GitHub Repositories
- [explanare/verbatim-memorization](https://github.com/explanare/verbatim-memorization) — Tools and experiments for injecting + detecting verbatim memorization in open LLMs (Pythia, etc.).
- [ftramer/LM_Memorization](https://github.com/ftramer/LM_Memorization) — Implements sampling-based extraction and memorization metrics (e.g. exposure, entropy).
- [sisinflab/LLM-MemoryInspector](https://github.com/sisinflab/LLM-MemoryInspector) — Probes whether LLMs retain structured records using controlled prompting.

**Comment by Garrafao** [comment]
Thanks.

> To test whether modern language models (TinyLLaMA and Flan-T5) show signs of training data contamination or leakage by reproducing sentences from the sdewac-v3.tagged corpus.

We are not interested in whether sdewac-v3.tagged corpus was part of the training data of these models. We are interested if the test data we use was part of the training data of models. Please adjust your objective and methodology. I would prefer if you could specify this in a minimal way, please not too much text or preliminary experiments. Please also to a short literature review and search for data leakage detection strategies and summarize them here. If you intend to use LLMs to summarize papers here, this is not a good approach as you will have to understand and implement things yourself.

Please use the same models as @prasanna9481 . We need to test leakage possibility on the models we use for prompting.

You already seem to have done some experiments. Please discuss such steps with me beforehand in order to make sure we align. As a reminder, I had only asked to define a set of experiments for the beginning.

Please adjust the above by Monday evening.

**Comment by divyansh-kaushik1** [comment]
Hello @Garrafao,

Thank you for the feedback. I want to ask a doubt. Aren't we interested in the sdewac-v3.tagged data(whole) leakage? I thought if the main data was checked okay, then the test data which is subpart of it is also good.

**Comment by Garrafao** [comment]
> Aren't we interested in the sdewac-v3.tagged data(whole) leakage?

This is only the corpus without gold labels, right? Why would we be interested in that?

> I thought if the main data was checked okay, then the test data which is subpart of it is also good.

What exactly is main data here and why should test be a subpart of it?

**Comment by divyansh-kaushik1** [comment]
Thank you for your response, I was thinking that as test_set.csv is a processed data form with labels from the sdewac dataset. So, if the original dataset in base form is leakage free then that would be good enough for both the test, and also highlight and give more confidence in zero shot prediction by prompting on the same data.

> What exactly is main data here and why should test be a subpart of it?

I was referring to the sdewac dataset as the main data.

**Comment by Garrafao** [comment]
> Thank you for your response, I was thinking that as test_set.csv is a processed data form with labels from the sdewac dataset.

Do we have such data anywhere? If yes, could you link it? If not, please give an example of a row in this data.

> > What exactly is main data here and why should test be a subpart of it?
> 
> I was referring to the sdewac dataset as the main data.

sdewac is not what we are concerned about regarding leakage.

**Comment by CastIehard** [comment]
I think you two are talking past each other.

@divyansh-kaushik1  The test set only contains words and their grammaticalization levels, and it is publicly available for download (as we also did). For us, it doesn’t matter whether the SDEWAC corpus was part of the model’s training data, because we are not asking the model to complete sentences.

But be careful: Our goal with the leakage check is **not** to test whether the LLM is “smart enough” to infer grammaticalization levels, but rather to check if it already knows this specific test set. If the LLM has seen the test set before, it might classify correctly without actually understanding grammaticalization.

So the real question is only whether the test set itself has leaked, not whether SDEWAC in general was part of training.

The testset can be found here:
[Processing/data/keyword_groundtruth.csv](../blob/main/Processing/data/keyword_groundtruth.csv)

_@divyansh-kaushik1 And please, do not add commits to a pull request I have already made—make your own PR instead. Your leakage detection code was merged as part of my data preprocessing PR, which makes the history confusing. Keeping these separate would be much clearer. Thanks <3_

**Comment by divyansh-kaushik1** [comment]
@CastIehard  Thank you for the response. I will be mindful of the next commit. I understood what the Professor meant after his last comment. I am currently working on it.
Thank you for your comment it reinforced my understanding of the task to do.

**Comment by Garrafao** [comment]
@divyansh-kaushik1 For your understanding of the topic and for planning our experiments, it may make sense for you to do a literature search for the term "data leakage llm" and similar, and to  summarize your findings here.

**Comment by divyansh-kaushik1** [comment]
## Leakage Detection Methods for Language Models (FlanT5 and TinyLlama Case Study)

After looking at other papers, this leakage test contains four independent experiments designed to detect potential **training data leakage** or **memorization** in large language models (LLMs). The experiments use the **FLAN-T5 Base** and **TinyLlama-1.1B-Chat-v1.0** model to evaluate its behavior on a custom dataset of German prepositions labeled with grammaticalization scores (test_set.csv) which is publicly avaialble for the download.

Each method explores a different aspect of model memorization or generalization.

---

## Method 1: Partial Keyword Recall

### Objective
Determine whether the model can reconstruct a **complete keyword** (e.g., a German preposition) from only its **first letter**, suggesting memorization or contamination.

### Procedure
- For each keyword in the test set, generate a prompt like `[v___]`.
- Ask the model to complete the word.
- Check if the full keyword appears in the model’s output.

### Metric
- **Recovery Rate**: Percentage of correct reconstructions of real keywords.

---

## Method 2: Real vs Fake Keyword Completion

### Objective
Compare the model’s ability to complete **real** versus **fake** (non-existent) prepositions, to expose bias toward memorized items.

### Procedure
- Use the same partial prompt strategy as Method 1.
- Run it twice:
  - Once on **real** prepositions from the test set.
  - Once on **fake**, plausible-looking but invented words.
- Compare success rates.

### Metric
- **Recovery Rate (Real)** vs. **Recovery Rate (Fake)**
- **Difference Score** = Real − Fake

---

## Method 3: Score-Based Keyword Prediction

### Objective
Evaluate the model's ability to generate **new prepositions** associated with a given **grammaticalization score**, suggesting category-level memorization.

### Procedure
- Group known keywords by score (1–4).
- Randomly sample 5 examples from each group.
- Prompt:  
  _"Here are some prepositions with score 2: [list]. Suggest another:"_
- Check if the model’s suggestion matches another keyword with the same score.

### Metric
- **Keyword Match Rate**: Fraction of predictions that match actual keywords of the same score.

---

## Method 4: Predict Score from Keyword

### Objective
Test whether the model can **assign the correct grammaticalization score (1–4)** to a given preposition — either from memorization or learned generalization.

### Procedure
- Prompt:  
  _"What is the grammaticalization score of the preposition ‘hinter’? Respond with a number from 1 (very low) to 4 (very high). "_
- Compare the model's output to the gold-standard label.

### Metrics
- **Per-Score Accuracy**: Accuracy broken down by class.
- **Overall Accuracy**

---

### Results
- The evaluation report for both of the model over these 4 is saved and uploaded in the repository.
- Both of the model did not any data leakage.

<img width="490" height="390" alt="Image" src="https://github.com/user-attachments/assets/8c5ccf92-100b-466a-b805-8dbef8e3605e" />

<img width="590" height="390" alt="Image" src="https://github.com/user-attachments/assets/2bfe5da6-38a5-42cc-a30b-e6a3027715f8" />

<img width="490" height="390" alt="Image" src="https://github.com/user-attachments/assets/a7bc2462-69df-4825-8042-98da9060088b" />

---
## References

| Method | Related Research |
|--------|------------------|
| **Method 1**<br>Partial Keyword Recall | Hayes et al. (2024) - [Measuring memorization in language models via probabilistic extraction](https://arxiv.org/html/2410.19482v2#:~:text=this%20issue%3A%20split%20a%20training,with%20respect%20to%20a%20single)  <br> Carlini et al. (2021) - [Extracting Training Data from Large Language Models](https://www.usenix.org/conference/usenixsecurity21/presentation/carlini-extracting#:~:text=It%20has%20become%20common%20to,by%20querying%20the%20language%20model) |
| **Method 2**<br>Real vs. Fake Completion | Wang et al. (2025) - [CON-RECALL](https://aclanthology.org/2025.coling-main.68.pdf#:~:text=Member%20prefixes%20boost%20log,models%E2%80%99%20funda%02mental%20tendency%20to%20generate)  <br> Xie et al. (2024) - [ReCaLL](https://arxiv.org/abs/2406.15968#:~:text=Conditional%20Log,Moreover%2C%20we%20conduct%20an%20in) |
| **Method 3**<br>Score-Based Keyword Generation | Yasunaga et al. (2024) - [Knowledge-Augmented Prompting](https://arxiv.org/abs/2310.01714#:~:text=%3E%20Abstract%3AChain,it%20obviates%20the%20need%20for)  <br> Brown et al. (2020) - [GPT-3 Few-Shot](https://arxiv.org/abs/2005.14165) |
| **Method 4**<br>Score Prediction from Keyword | Petroni et al. (2019) - [LAMA Probe](https://arxiv.org/abs/1909.01066)  <br> Yasunaga et al. (2024) - [Knowledge-Augmented Prompting](https://arxiv.org/abs/2310.01714#:~:text=%3E%20Abstract%3AChain,it%20obviates%20the%20need%20for) |

**Comment by CastIehard** [comment]
Thanks @divyansh-kaushik1 for the overview. I have a few critical observations regarding the methods:

Method 1 raises some fundamental concerns. I don't see how this method tells us anything about the model's memory of the grammaticalization scores. The ability to complete a word from its first letter only shows that the LLM has seen the word before — which we already know. What matters for leakage detection is whether the model has seen the test set and especially the gold-standard labels. This method seems disconnected from that goal? Or did I misunderstand what you were doing?

Method 3, on the other hand, seems more relevant and well-aligned with what we're trying to detect. However, I’d suggest expanding the setup: introduce more diverse prompts, like: "Predict Grammaticalization Level as in Dominik et al. Paper... that were stored in the ground truth testset...."

Method 4 seems problematic from a leakage detection standpoint. It doesn't help distinguish whether the model predicts the grammaticalization score through memorization (i.e., data leakage) or logical generalization based on learned linguistic patterns. Without such disambiguation, the results risk being misleading — it could just replicate what a baseline prompt from @prasanna9481 like “Predict the grammaticalization score” would already output, without proving anything about leakage. This method doesn't bring added value for the specific goal of leakage assessment.

**Comment by divyansh-kaushik1** [comment]
Thank you for detailed feedback @CastIehard.

For Method 1, the target is not to test for memorization of grammaticalization scores. It is the general assessment of the models. To check, if the model see these specific prepositions in training? And more importantly, does it target them when only minimal input is given? If the model can recover full keywords from just one letter, that suggests verbatim recall, which is a precursor to label leakage. This mirrors methods used in Carlini et al. (2021) and Hayes et al. (2024) for detecting contamination via prefix completions.

The Method 2 handles the Generalization case of the word seen before and checks in further and builds on method 1. It does so by contrasting completions of real vs fake keywords, we shift from detecting memorization to selective memorization. In that sense, Method 2 acts as filter check: it helps validate whether recall observed in Method 1 reflects actual leakage or just fluent guessing. 

You have given a great suggestion for Method 3, to enrich prompts with dataset references or citations (e.g., “as in Dominik et al.”). That could sharpen the task and better simulate leakage conditions.

The concern fair, thanks for pointing this out. This one is more ambiguous, I agree as it blurs the line between memorization and generalization. But we can see it as a zero-shot probe, if the model assigns correct scores with no examples, it could indicate leakage, especially on rare or low-frequency items. It’s not definitive alone, but helps triangulate with the other methods.

These were the thought process that went into these methods. Further ideas and suggestion will be very much appreciated.  Once again thank you.
